apps/prairielearn/src/pages/authCallbackLti/authCallbackLti.js

```diff
@@ -15,202 +13,168 @@ const { config } = require('../../lib/config');
+      authn_provider_name: 'LTI',
+    };
+    const pl_authn = generateSignedToken(tokenData, config.secretKey);
+    res.cookie('pl_authn', pl_authn, {
+      maxAge: config.authnCookieMaxAgeMilliseconds,
+      httpOnly: true,
+      secure: shouldSecureCookie(req),
+    });
+
+    // Dual-write information to the session so that we can start reading
+    // it instead of the cookie in the future.
+    req.session.user_id = userResult.rows[0].user_id;
+    req.session.authn_provider_name = 'LTI';
+
+    const linkResult = await sqldb.queryOneRowAsync(sql.upsert_current_link, {
+      course_instance_id: ltiresult.course_instance_id,
+      context_id: parameters.context_id,
+      resource_link_id: parameters.resource_link_id,
+      resource_link_title: parameters.resource_link_title || '',
+      resource_link_description: parameters.resource_link_description || '',
+    });
+
+    // Do we have an assessment linked to this resource_link_id?
+    if (linkResult.rows[0].assessment_id) {
+      if ('lis_result_sourcedid' in parameters) {
+        // Save outcomes here
+        await sqldb.queryAsync(sql.upsert_outcome, {
+          user_id: tokenData.user_id,
+          assessment_id: linkResult.rows[0].assessment_id,
+          lis_result_sourcedid: parameters.lis_result_sourcedid,
+          lis_outcome_service_url: parameters.lis_outcome_service_url,
+          lti_credential_id: ltiresult.id,
+        });
+      }
 
-      var genSignature = oauthSignature.generate(
-        'POST',
-        ltiRedirectUrl,
-        parameters,
-        ltiresult.secret,
-        undefined,
-        { encodeSignature: false },
+      res.redirect(
+        `${res.locals.urlPrefix}/course_instance/${ltiresult.course_instance_id}/assessment/${linkResult.rows[0].assessment_id}/`,
       );
-      if (genSignature !== signature) {
-        return next(error.make(500, 'Invalid signature'));
-      }
+    } else {
+      // No linked assessment
 
-      // Check oauth_timestamp within N seconds of now (3000 suggested)
-      var timeDiff = Math.abs(Math.floor(Date.now() / 1000) - parameters.oauth_timestamp);
-      if (timeDiff > timeTolerance) {
-        return next(error.make(500, 'Invalid timestamp'));
-      }
+      const instructorResult = await sqldb.callAsync('users_is_instructor_in_course_instance', [
+        tokenData.user_id,
+        ltiresult.course_instance_id,
+      ]);
 
-      // Check nonce hasn't been used by that consumer_key in that timeframe
-      // https://oauth.net/core/1.0/#nonce
-      var nonceReused = false;
-      var nonceKey = 'authCallbackLti:' + parameters.oauth_timestamp + ':' + parameters.oauth_nonce;
-      util.callbackify(cache.get)(nonceKey, (err, val) => {
-        if (ERR(err, next)) return;
-        if (val) {
-          nonceReused = true;
-        } else {
-          cache.set(nonceKey, true, timeTolerance * 1000);
-        }
-      });
-      if (nonceReused) return next(error.make(500, 'Invalid nonce reuse'));
```

Comment:
This is what I was aiming to fix. The code as written is nonsensical: it starts an async task to fetch the nonce from the cache but then synchronously reads `nonceReused`, which will always be `false` at that point.